doc/hdup.1
.\" @(#)hdup.1 1.7.0 20-Mar-2003 OF; 
.TH hdup 1 "18 Mar 2003"
.SH NAME
hdup \- harddisk duplicator/harddisk backupper - backup to harddisk
.SH SYNOPSIS
.B hdup
[
.IR OPTION
]
.IR SCHEME 
.IR HOST 
[
.IR @USER@REMOTEHOST
]
\fI(1st format)\fR
.br
.B hdup
[
.IR OPTION
]
.IR restore
.IR HOST
.IR DATE
.IR DIRECTORY
[
.IR @USER@REMOTEHOST
]
\fI(2nd format)\fR

.SH DESCRIPTION
\fBHdup\fR is used to backup a filesystem. Features include:
.PD 1
.TP
.B o
incremental backups: monthly, weekly and daily dumps,
.TP
.B o
encryption of the archive (via \fBmcrypt\fR or \fBGPG\fR),
.TP
.B o
compression of the archive (bzip/gzip/lzop/none),
.TP
.B o
possibility to transfer the archive to a remote host, 
.TP
.B o
possibility to restore the archive from a remote host, 
.TP
.B o
ability to split up archives,
.TP
.B o
no obscure archive format (it is a normal compressed tar file), and
.TP
.B o
simple to use.
.PP
The behaviour of \fBhdup\fR is controlled by its configuration file (see
\fBhdup.conf(5)\fR). Internally \fBhdup\fR uses \fBGNU tar\fR to actually create
the backups.

.SS First format
When using the 1st format \fBhdup\fR performs a backup. Remember: \fBhdup\fR
pushes a backup \fIfrom\fR the localhost \fIto\fR the remote host.

The sort of backup is specified by \fISCHEME\fR:
.TP 
\fBmonthly\fR
Make a full (null) dump of the filesystem.
.TP 
\fBweekly\fR
Make an incremental dump of the filesystem relative to the latest
monthly dump. If \fBhdup\fR cannot find a monthly dump it will complain, unless
\fIalways backup\fR is on. Then a monthly dump will be performed.
.TP 
\fBdaily\fR
Make an incremental dump of the filesystem relative to the latest
weekly dump. If \fBhdup\fR cannot find a weekly dump it will complain, unless
\fIalways backup\fR is on. Then a weekly dump will be performed.

.PP
\fIHOST\fR is the host of which \fBhdup\fR should perform the backup. This
should match a '[HOST]' statement in the configuration file. The
directories of that host (specified with 'dir = \fIdir1\fR, \fIdir2\fR', ...) will
be backed up to the directory specified with 'archive dir = \fIdir\fR'. 
If \fIHOST\fR is not found no backup will be made.

.PP
\fI@USER@REMOTEHOST\fR is the host to which the archive should be
transfered. This must include the user name. E.g \fI@miekg@elektron.atoom.net. 
It is illegal to specify the colon ':'. \fBhdup\fR must be
present on the remote host. The location of this remote \fBhdup\fR is specified using
\fIremote hdup\fR.

Any program capable of transferring files can used for this purpose. Currently
tested is \fBssh\fR. Other programs like \fBrsync\fR (not tested) may also work. Any program
with the following characteristics will do:
.TP
.B o
must be usable as a filter (read from stdin, write to stdout),
.TP
.B o
must support \fIuser@remotehost\fR syntax.

.SS Second format
When using the 2nd format a previous backed up filesystem is restored.
Remember: \fBhdup\fR pushes a restore \fIfrom\fR the remote \fIto\fR 
the local host. This is opposite from the backup operation!

.PP
\fIHOST\fR is the host who's archives should be restored.

.PP
\fIDATE\fR everything up to this date will be restored. \fBhdup\fR will
look for the most recent monthly archive, then the most recent weekly
and finally for the daily to pad up to date given. The \fIDATE\fR can
either be specified as DD-MM-YYYY (date spec = default), as YYYY-MM-DD
(date spec = iso) or as MM-DD-YYYY (date spec = american). This is controlled
from the configuration file. 

A special date is 'static' which instructs
\fBhdup\fR to look in the 'static' directory. This is used when 'no history =
yes'. This is dangerous to use because \fBhdup\fR will overwrite the old backup
file with the new one. When your system crashes during the overwrite you have
no backup at all! Only use this when you \fIreally\fR don't have room for two
monthly backups.

Another special date is 'today' which instructs \fBhdup\fR to use the current date.

.PP
\fIDIRECTORY\fR tells \fBhdup\fR to which directory the archive should
be untarred to. Be very careful when running \fBhdup\fR as root and
specifying '/' as the directory. Version 1.4 and above refuses to restore to '/'.
This can be overridden by specifying 'force = on/yes' in the configuration.

.PP
\fI@USER@REMOTEHOST\fR is the host to which the archive should be restored.
This must include the user name. E.g \fI@miekg@elektron.atoom.net.
As as version 1.6.6 it is illegal to specify the colon ':'. On the remosthost
and in \fIDIRECTORY\fR the archive is restored. \fBhdup\fR must be present 
on the remote host.

.SS Status message
When \fBhdup\fR is finished with its current operation it will print
an overview message:

.TS
tab ($);
l l.
Hdup version.:  1.6.6

Host.........:  elektron
Date.........:  2003-02-02
Scheme.......:  monthly
Archive......:  elektron.2003-02-02.monthly.tar.gz
Encryption...:  no
Archive size.:  257k
Elapsed......:  0:01:27
Status.......:  successfully performed backup
.TE
.PP
Which can be mailed to you via cron.

.SH OPTIONS
.TP
\fB\-c, --config=\fIconfig\fR
Location of the configuration file. The default location of \fBhdup\fR's
configuration file is \fI/etc/hdup/hdup.conf\fR.
.TP
\fB\-s, --specific=\fIfile\fR
Restore a specific file from an archive. \fIfile\fR must be the \fBfull\fI path
to the file, relative paths will not work.
.TP
\fB\-i, --ignore-tar
Ignore tar errors when restoring.
.TP
\fB\-I, --ignore-conf
Ignore errors in the configuration file.
.TP
\fB\-P, --patched_tar
Tar is patched so that it can handle \fI--no-recursion\fR,
\fI--listed-incremental\fR and \fI--files-from\fR together. This options
enables two things in hdup; 1) directory info is written to 'filelist' and
2) \fI--no-recursion\fR is given to \fBtar\fR.

This solves the bug whereby \fBhdup\fR wouldn't include directory
information in the archives.
.TP
\fB\-d, --dryrun
Do a dryrun - don't do anything with the filesystem
.TP
\fB\-q, --quiet
Suppress the output of the subprocesses (like 'tar' and 'ssh').
.TP
\fB\-q \-q, --quiet --quiet
Suppress the logging output from \fBhdup\fR.
.TP
\fB\-q \-q \-q, --quiet --quiet --quiet
No logging at all. Even no overview message.
.TP
\fB\-V
Be more verbose.
.TP
\fB\-V \-V
Be even more verbose. This will show which files are backed up by
\fBhdup\fR IF you also supply the -D option.
.TP
\fB\-h, --help
A help message.
.TP
\fB\-v, --version
Show the version of hdup.
.TP
\fB\-D, --debug
Show a lot of information which can aid debugging. 

.PP
The -V and -q options do \fInot\fR effect each other. A '-qqq -VV' option list
will mean that \fBhdup\fR will show what is run, but nothing else (no overview message
and no warning nor errors).

.SH ENCRYPTION
\fBhdup\fR can encrypt the archives, \fBmcrypt\fR is used for the actual
encryption. As of version 1.6.25 \fBGPG\fR can also be used to encrypt
the archive. Note that currently remotely restoring a GPG encrypted archives
is not working.

.SS mcrypt
With \fBmcrypt --list\fR you get a list of the algorithms
mcrypt supports:

.TS
tab (@); 
l l.
\.\.\.
\fIserpent\fR (32): cbc cfb ctr ecb ncfb ofb nofb 
\fIwake\fR (32): stream 
\fIloki97\fR (32): cbc cfb ctr ecb ncfb ofb nofb 
\fIrijndael-128\fR (32): cbc cfb ctr ecb ncfb ofb nofb 
\fIrijndael-192\fR (32): cbc cfb ctr ecb ncfb ofb nofb 
\.\.\.
.TE
.PP 
If you want to use \fIloki97\fR you specify \fBalgorithm = loki97\fR in
hdup.conf. The same goes for all the other algorithms.

.SS GPG
To use \fBGPG\fR the following is needed. Set \fBalgorithm = gpg\fR and
\fBkey = user ID of key\fR. In my case I've created a GPG key with user ID
of 'miekg' so I use:
.br
\fBalgorithm = gpg\fR
.br
\fBkey = miekg\fR
.br
The \fBkey\fR is supplied to \fBgpg\fR with the \fI-r\fR argument. See the manpage 
of \fBgpg\fR for more details.
.PP
When restoring a GPG encrypted archive you will be prompted to unlock your
private key.

.SH AUTHOR
Written by Miek Gieben. Wouter van Gils helped a lot with testing pre-release
versions. User feedback is appreciated.

.SH REPORTING BUGS
Report bugs to <hdup-user@miek.nl>. 

.SH BUGS
The %a expansion is not always the same in the prerun and postrun scripts (when
using encryption).

.SH LIMITATIONS
Under Linux kernel version 2.2 the archive size cannot exceed two (2) Gigabyte.
If you need larger archives sizes you should upgrade your kernel. You can
however solve this by using \fIchunk size\fR. Just define your maximum allowed
size, something like \fIchunk size = 1800M\fR and you're set.
.PP
If you encrypt archives and want to restore them, you are forced to use one
encryption scheme for all the backups. \fBhdup\fR does not store the key
and algorithm with the archive, thus it is impossible to restore archives that
use different keys and algorithms.

.SH COPYRIGHT
Copyright (C) 2001-2005 Miek Gieben. This is free software. There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.

.SH SEE ALSO
\fBhdup.conf(5)\fR for information about \fBhdup\fR's configuration file.


contrib/hdup.sh
#!/bin/bash

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABLILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details. You should have received a
# copy of the GNU General Public License along with this program; if
# not, write to the Free Software Foundation, Inc, 59 Temple Place -
# Suite 330, Boston, MA 02111-1307, USA.

# On Debian systems, the complete text of the GNU General Public
# License can be found in /usr/share/common-licenses/GPL file.

# For questions please contact: Juraj.Kubelka@email.cz

function snapshot_name {
# $1 is name of LVM Logical Volume
# Return name of Snapshot for this LVM Logical Volume
echo "${1##*/}_SNAPSHOT"
}

function snapshot_name_long {
# $1 is name of LVM Logical Volume
# Return name of Snapshot for this LVM Logical Volume
#        - in logn form (/dev/.../...)
echo "${1%/*}/$(snapshot_name "$1")"
}

function snapshot_directory {
# $1 is name of LVM Logical Volume
# Return name of directory where you can mount its snapshot
echo "/$(snapshot_name "$1")"
}

function create_snapshot {
# $1 is name of LVM Logical Volume for what you want to create snapshot
#    ... required
# $2 is name of directory where you want to mount this snapshot
# $3 is size of snapshot you want to create

if [ -z "$1" ] ; then
    echo "Specify LVM Logical Volume for what you want to create snapshot" >& 2
fi

SNAP_NAME="$(snapshot_name "$1")"
SNAP_NAME_LONG="$(snapshot_name_long "$1")"
SNAP_MOUNT_POINT="${2:-$(snapshot_directory "$1")}"
SNAP_SIZE=${3:-4G}

lvcreate -L"$SNAP_SIZE" -s -n "$SNAP_NAME" "$1" &&
mkdir "$SNAP_MOUNT_POINT" &&
mount -o ro "$SNAP_NAME_LONG" "$SNAP_MOUNT_POINT"

return $?
} # create_snapshot

function remove_snapshot {
# $1 is name of LVM Logical Volume you want to umount and remove its Snapshot
#    ... required
# $2 is name of directory where this snapshot is mounted

if [ -z "$1" ] ; then
    echo "Specify LVM Logical Volume you want to remove its snapshot" >& 2
fi

SNAP_NAME="$(snapshot_name "$1")"
SNAP_NAME_LONG="$(snapshot_name_long "$1")"
SNAP_MOUNT_POINT="${2:-$(snapshot_directory "$1")}"

umount "$SNAP_MOUNT_POINT"
rmdir "$SNAP_MOUNT_POINT"
lvremove -f "$SNAP_NAME_LONG"
} # remove_snapshot


LOCK_FILE="/tmp/hdup.lock"
    
function unlock_backups {

    rm -f $LOCK_FILE
} # unlock_backups

function lock_backups {
# $1 is scheme you want to use (daily, weekly, monthly)

    lockfile -0 -r1 $LOCK_FILE && \
	trap 'unlock_backups ; exit 1' 1 2 3 15 || \
	( 
	echo "HDUP is runnig (scheme: $(cat $LOCK_FILE))." >& 2
	return 1
	) || return 1

    chmod u+w $LOCK_FILE &&
    echo -n "$1" > $LOCK_FILE &&
    chmod u-w $LOCK_FILE ||
    ( 
	unlock_backups
	return 1
    ) || return 1
} # lock_backups

function backup_host {
# $1 is scheme you want to use (daily, weekly, monthly)
# $2 is name of host you want to backup

    echo ""
    echo "+++ Start hdup backup: $1, $2 +++" 
    nice -n 2 hdup -q -q "$1" "$2"
    RET=$?
    echo "+++  Done hdup backup: $1, $2 +++"
    echo ""
    return $RET
 } # backup_host

function backup_all_hosts {
# $1 is scheme you want to use (daily, weekly, monthly)
    
    for i in $(cat /etc/hdup/hdup.conf \
	| grep -e '^\[[^[]*\]' \
	| grep -v global \
	| sed -e 's/\[//g' -e 's/\]//g')
      do 
      backup_host "$1" "$i" || \
	  return 1
    done

    run_post_backups "$1"
} # backup_all_hosts

function backup_all_hosts_with_lock {
# $1 is scheme you want to use (daily, weekly, monthly)
    
    if lock_backups "$1" ; then
	backup_all_hosts "$1" &&
	unlock_backups "$1" ||
	(
	    echo "An error during backup." >& 2
	    echo "  Write \`$0 unlock\` before new start." >& 2
	    return 1
	) || return 1
    fi
} # backup_all_hosts_with_lock

POST_BACKUPS_ADR="/etc/hdup/post-backup.d"

function run_post_backups {
# $1 is scheme you want to use (daily, weekly, monthly)

    export HDUP_SCHEME="$1"

    [ ! -d "$POST_BACKUPS_ADR" ] && return 0

    echo ""
    echo "+++ Start post backups: $1 +++"

    ls "$POST_BACKUPS_ADR" \
	| \
	while read backup_script
      do
      [ ! -x "$POST_BACKUPS_ADR/$backup_script" ] && continue
      
      echo -n "Script ($backup_script): "
      "$POST_BACKUPS_ADR/$backup_script" && \
	  echo "done." || \
	  echo "An ERROR!!!"
    done
    
    echo "+++ Done post backups: $1 +++"
    echo ""
} # run_post_backups

HDUP_OUTPUT="/tmp/hdup.output.$$.txt"

function backup_and_send_email {
# Backup and send email with backup results.
# $1 is scheme you want to use (daily, weekly, monthly)
    
    HDUP_START_DATE="$(date)"
    backup_all_hosts_with_lock $* 2>&1 | tee "$HDUP_OUTPUT"
    HDUP_END_DATE="$(date)"

    (
	echo "HDUP start time: $HDUP_START_DATE."
	echo "HDUP end time  : $HDUP_END_DATE."
	echo ""
	cat "$HDUP_OUTPUT"
    ) \
	| mail -s "HDUP backup, scheme $1." root@localhost

    rm -f "$HDUP_OUTPUT"
} # backup_and_send_email

function help {

cat <<EOF

Wrapper for hdup application 
 (see http://miek.nl/projects/hdup2/index.html).

Questions/remarks should go to: Juraj.Kubelka@email.cz

USAGE: ${0##*/} <command> [parameters]

Commands for backup:

  backup <scheme>
        - backup all hosts
  backup-by-email <scheme>
        - backup all hosts and send report by email
          Email is sent to root@localhost.
          There is used screen application.
  backup-by-email-no-screen <scheme>
        - backup all hosts and send report by email
          There isn't used screen application.

  parameters:
     scheme = { daily | weekly | monthly }
        - see man hdup

Commands for support backup:

  lock  
        - lock backup process
          Preserve before run backup process. It can be used for
          disable automatic run backup process.

  unlock
        - unlock backup process

  run-post-backups [scheme]
        - rerun post backup process manually
          Usually it is run after all hosts are backup.
          Scripts are located in ${POST_BACKUPS_ADR} directory.

          When it is used automatically after backup process, variable
          HDUP_SCHEME is set to scheme { daily | weekly | monthly }.
          So it can be used inside scripts.

  create-snapshot <LVM-name> [mount-point] [Snapshot-size]
        - Create snapshot of LVM-name (see man lvcreate). Size will be
          set to Snapshot-size. Mount point will be mount-point.

  remove-snapshot <LVM-name> [mount-point]
        - Remove snapshot of LVM-name.

          When I've tested it last time (about February 2005), there
          were some kernel problems with allocating memory for
          "snapshot table". So now I don't use it.

          Idea:
          Write in hdup config file (/etc/hdup/hdup.conf) scheme:

          [scheme-usr-local]
          dir = /SYSTEM-USR-SNAPSHOT/local
          prerun = ${0} create-snapshot /dev/SYSTEM/USR /SYSTEM-USR-SNAPSHOT
          postrun = ${0} remove-snapshot /dev/SYSTEM/USR /SYSTEM-USR-SNAPSHOT

EOF

} # help

# --------------------------------------------

if [ "$1" = "create-snapshot" ] ; then
    shift
    create_snapshot $*
elif [ "$1" = "remove-snapshot" ] ; then
    shift
    remove_snapshot $*

elif [ "$1" = "lock" ] ; then
    lock_backups testing
elif [ "$1" = "unlock" ] ; then
    unlock_backups 

elif [ "$1" = "backup" ] ; then
    shift
    backup_all_hosts_with_lock $*
elif [ "$1" = "backup-by-email" ] ; then
    shift
    screen -t hdup.backup."$1" bash -c "$0 backup-by-email-no-screen $*"
elif [ "$1" = "backup-by-email-no-screen" ] ; then
    shift
    backup_and_send_email $*

elif [ "$1" = "run-post-backups" ] ; then
    shift
    run_post_backups $*
elif [ \( "$1" = "help" \) -o \( "$1" = "-h" \) -o \( "$1" = "--help" \) ] ; then
    help
fi
